### PR TITLE
alternative approach to removing ServerOptionsBuilderOptions

### PR DESCRIPTION
**Pull request by peterbom**
Rather than changing the behaviour of the builder itself, I wanted to see if we can change the CertificateStore implementation so that it doesn't block testing.

### CODE REVIEW

File: src/Microsoft.Azure.Batch.SoftwareEntitlement.Common/CertificateStore.cs [hunk 1]
```diff
@@ -6,10 +6,17 @@
 
 namespace Microsoft.Azure.Batch.SoftwareEntitlement.Common
 {
+    public interface ICertificateStore
+    {
+        IEnumerable<X509Certificate2> FindAll();
+
+        Errorable<X509Certificate2> FindByThumbprint(string purpose, CertificateThumbprint thumbprint);
+    }
+
```
Comment (theunrepentantgeek) [inline src/Microsoft.Azure.Batch.SoftwareEntitlement.Common/CertificateStore.cs:15]: Needs documentation comments - copy them from the implementation

File: src/sestest/ServerOptionBuilder.cs [hunk 2]
```diff
@@ -16,38 +14,29 @@ public class ServerOptionBuilder
...
         /// command line by the user
         /// </summary>
         /// <param name="commandLine">Command line parameters supplied by the user.</param>
-        /// <param name="options">Options for configuring validation.</param>
         /// <returns>Either a usable (and completely valid) <see cref="ServerOptions"/> or a set 
         /// of errors.</returns>
-        public static Errorable<ServerOptions> Build(
-            ServerCommandLine commandLine,
-            ServerOptionBuilderOptions options = ServerOptionBuilderOptions.None)
+        public static Errorable<ServerOptions> Build(ServerCommandLine commandLine, ICertificateStore certificateStore)
```
Comment (theunrepentantgeek) [inline src/sestest/ServerOptionBuilder.cs:26]: I think we should get rid of the static method - it was there to hide `ServerOptionBuilderOptions.None`

File: src/sestest/ServerOptionBuilder.cs [hunk 3]
```diff
@@ -16,38 +14,29 @@ public class ServerOptionBuilder
...
 
         /// <summary>
         /// Initializes a new instance of the <see cref="ServerOptionBuilder"/> class
         /// </summary>
         /// <param name="commandLine">Options provided on the command line.</param>
-        /// <param name="options">Options for configuring validation.</param>
-        private ServerOptionBuilder(
-            ServerCommandLine commandLine,
-            ServerOptionBuilderOptions options)
+        private ServerOptionBuilder(ServerCommandLine commandLine, ICertificateStore certificateStore)
```
Comment (theunrepentantgeek) [inline src/sestest/ServerOptionBuilder.cs:36]: So this needs to become public.

File: tests/Microsoft.Azure.Batch.SoftwareEntitlement.Tests/FakeCertificateStore.cs [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+using System.Linq;
+using System.Security.Cryptography.X509Certificates;
+using Microsoft.Azure.Batch.SoftwareEntitlement.Common;
+
+namespace Microsoft.Azure.Batch.SoftwareEntitlement.Tests
+{
+    public class FakeCertificateStore : ICertificateStore
+    {
+        private readonly string _notFoundError;
+        private readonly Dictionary<string, X509Certificate2> _certLookup;
```
Comment (theunrepentantgeek) [inline tests/Microsoft.Azure.Batch.SoftwareEntitlement.Tests/FakeCertificateStore.cs:11]: Make this `Dictionary<CertificateThumbprint, X509Certificate2>` - the `CertificateThumbprint` struct implements `.Equals(object)` and `.GetHashCode()` properly, so it should work as a key for the Dictionary.

File: tests/Microsoft.Azure.Batch.SoftwareEntitlement.Tests/FakeCertificateStore.cs [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+        {
+            return _certLookup.Values;
+        }
+
+        public Errorable<X509Certificate2> FindByThumbprint(string purpose, CertificateThumbprint thumbprint)
+        {
+            return _certLookup.ContainsKey(thumbprint.ToString())
+                ? Errorable.Success(_certLookup[thumbprint.ToString()])
+                : Errorable.Failure<X509Certificate2>(_notFoundError);
+        }
```
Comment (theunrepentantgeek) [inline tests/Microsoft.Azure.Batch.SoftwareEntitlement.Tests/FakeCertificateStore.cs:31]: This is traversing the dictionary's internal structures twice. Better this way:
```
if (_certLookup.TryGetValue(thumbprint out var certificate)) {
    return Errorable.Success(certificate);
}

return Errorable.Failure<X509Certificate2>(_notFoundError);
```

File: tests/Microsoft.Azure.Batch.SoftwareEntitlement.Tests/ServerOptionBuilderTests.cs [hunk 6]
```diff
@@ -9,48 +11,44 @@ namespace Microsoft.Azure.Batch.SoftwareEntitlement.Tests
...
-        //    property value set.
-        // To ensure the second part is not missed, we're unit testing both of these
-        // actions together. We want to test each property individually, but some
-        // properties are intended to be mandatory, and their absence will cause
-        // errors unrelated to the property being tested. To work around this we use
-        // a "permissive" option which changes the behaviour of the builder such that
-        // the enforcement of mandatory properties is bypassed.
-        private readonly ServerOptionBuilderOptions _permissiveOptions =
-            ServerOptionBuilderOptions.ConnectionThumbprintOptional;
+        private readonly ICertificateStore _certificateStoreWithConnectionCertificate = new FakeCertificateStore(
```
Comment (theunrepentantgeek) [inline tests/Microsoft.Azure.Batch.SoftwareEntitlement.Tests/ServerOptionBuilderTests.cs:25]: I think we could call this `_certificateStore`.

**Review by theunrepentantgeek — CHANGES_REQUESTED** [review]
You've convinced me - this is really clean. 👍 
A few minor bits of cleanup.

Submit this as a PR to the main repo, I'll merge it, and abandon my spike.

**Comment by peterbom** [comment]
Already merged as separate PR

**Pull request closed without merging**